### PR TITLE
Add machine + runtime identity to docker labels serialization

### LINKED ISSUE #4104: Serialize machine required fields as docker container labels
This is a very first step to achive workspace runtimes recovery.
Basically each time docker container is created machine servers are stored as labels, which 
allows to recover this data in future. The other machine mandatory data such as 
```
machine id
machine name
machine owner
workspace id
environment name
``` 
should be also serialized as container labels, proposed format is:

```
"che:env_name": "wksp-lz1b",
"che:id": "machinep91hshmqt17bi7lp",
"che:name": "dev-machine",
"che:owner": "che",
"che:workspace_id": "workspaceos6k0l3pz89ufi2z",
"che:server:22/tcp:ref": "ssh",
"che:server:4401/tcp:path": "/api",
"che:server:4401/tcp:protocol": "http",
"che:server:4401/tcp:ref": "wsagent",
"che:server:4411/tcp:protocol": "http",
"che:server:4411/tcp:ref": "terminal",
"che:server:8000:protocol": "http",
"che:server:8000:ref": "tomcat8-debug",
"che:server:8080:protocol": "http",
"che:server:8080:ref": "tomcat8",
"che:server:9876:protocol": "http",
"che:server:9876:ref": "codeserver"
```

Investigated by #3910

### PR DESCRIPTION
**Pull request by voievodin**
### What does this PR do?
Adds machine name + runtime identity + server configurations to docker labels serialization.

I tried to follow [key format recommendations](https://docs.docker.com/engine/userguide/labels-custom-metadata/), so the proposed format was a bit changed. 
Serialized labels
```
"org.eclipse.che.machine.name": "default",
"org.eclipse.che.workspace.env": "test",
"org.eclipse.che.workspace.id": "workspace28t00rschopreajr",
"org.eclipse.che.workspace.owner": "che"
"org.eclipse.che.server.exec-agent.port": "4412/tcp",
"org.eclipse.che.server.exec-agent.protocol": "http",
"org.eclipse.che.server.ssh.port": "22/tcp",
"org.eclipse.che.server.ssh.protocol": "ssh",
"org.eclipse.che.server.terminal.port": "4411/tcp",
"org.eclipse.che.server.terminal.protocol": "http",
"org.eclipse.che.server.wsagent.port": "4401/tcp",
"org.eclipse.che.server.wsagent.protocol": "http",
```

### What issues does this PR fix or reference?
Resolves #4104

### CODE REVIEW

**Review by sleshchenko — APPROVED** [review]

File: infrastructures/docker/src/main/java/org/eclipse/che/workspace/infrastructure/docker/DockerInternalRuntime.java [hunk 1]
```diff
@@ -194,68 +200,56 @@ protected void internalStop(Map<String, String> stopOptions) throws Infrastructu
...
+        startMachine(name, config);
+    }
 
-                DockerContainerConfig imageContainerConfig = normalizeSource(containerConfig, machineSource);
-                dockerMachine = serviceStarter.startService(dockerEnvironment.getNetwork(),
-                                                            name,
-                                                            imageContainerConfig,
-                                                            identity,
-                                                            isDev);
-            } catch (NotFoundException | SnapshotException | SourceNotFoundException e) {
```
Comment (garagatyi) [inline infrastructures/docker/src/main/java/org/eclipse/che/workspace/infrastructure/docker/DockerInternalRuntime.java:224]: Here is one tricky case. When SourceNotFoundException is thrown we should fall back to start from origin source. Please, add this workaround.

File: infrastructures/docker/src/main/java/org/eclipse/che/workspace/infrastructure/docker/Labels.java [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Helps to deal with docker container labels.
```
Comment (garagatyi) [inline infrastructures/docker/src/main/java/org/eclipse/che/workspace/infrastructure/docker/Labels.java]: Javadoc describe the purpose of this class as it has no connection to DockerMachine. Can you rephrase it to show the connection to labels related to a machine?

Comment (voievodin) [reply infrastructures/docker/src/main/java/org/eclipse/che/workspace/infrastructure/docker/Labels.java]: ok

File: infrastructures/docker/src/main/java/org/eclipse/che/workspace/infrastructure/docker/DockerInternalRuntime.java (hunk 1 above)
Comment (voievodin) [reply infrastructures/docker/src/main/java/org/eclipse/che/workspace/infrastructure/docker/DockerInternalRuntime.java:224]: Thanks, i missed that

**Comment by voievodin** [comment]
@garagatyi i fixed behaviour please review again

**Review by garagatyi — APPROVED** [review]

**Comment by garagatyi** [comment]
Forget to add in review: good refactoring!

**Comment by voievodin** [comment]
Thank you for the review guys!
